File: 01-SRP/CPP/SOLUTION/main.cpp
/**
* El código cumple con el SRP al separar la gestión de la entidad (Empleado) 
* de la gestión de la persistencia (EmpleadoRepositorio). Ahora, si el formato
* de guardado cambia (a base de datos, por ejemplo), solo debe modificarse 
* la clase EmpleadoRepositorio.
*/
#include <iostream>
#include <fstream>
#include <string>
#include <memory>

class Empleado 
{
    public:
        std::string nombre;
        double salario;

        Empleado(std::string n, double s) : nombre(n), salario(s) {}

        std::string getNombre() const { return nombre; }
        double getSalario() const { return salario; }
};

class EmpleadoRepositorio 
{
    public:
        void guardar(const std::shared_ptr<Empleado>& emp) 
        {
            std::ofstream file("empleados.txt", std::ios::app);
            file << emp->getNombre() << "," << emp->getSalario() << "\n";
            file.close();
            std::cout << "Empleado guardado en archivo.\n";
        }
};

int main()
{
    auto juan = std::make_shared<Empleado>("Ana", 30.5);
    auto repo = std::make_shared<EmpleadoRepositorio>();
    repo->guardar(juan);

    return 0;
}

File: 01-SRP/TYPESCRIPT/SOLUTION/main.ts
// 1. Importación Correcta: Usamos 'import' para módulos Node.js/CommonJS en TS.
//    También necesitamos instalar los tipos de Node.js (@types/node) para que esto funcione.
import * as fs from 'fs';

class Empleado 
{
    constructor(public nombre: string, public salario: number) {}

    getNombre(): string { return this.nombre; }
    getSalario(): number { return this.salario; }
}

class EmpleadoRepositorio 
{
    guardar(emp: Empleado): void
    {
        const fs = require('fs');
        fs.appendFileSync('empleados.txt', `${emp.getNombre()},${emp.getSalario()}\n`);
        console.log('Empleado guardado en archivo.');
    }
}

function main()
{
    let juan = new Empleado("Ana", 30.5);
    let repo = new EmpleadoRepositorio();
    repo.guardar(juan);
}

main();